Reordering clarification and update diagrams

peterbrightwell opened the pull request:
See #28 

There is still one TODO in 4.0. Behaviour.md that require a JSON example of re-entrant mapping. I suggest another PR for that.

peterbrightwell committed:
Some updates towards fixing #28 following call with GSB, CG, AB

peterbrightwell committed:
Resize to 35%, 150 dpi

peterbrightwell committed:
Resize 20%, dpi 300

peterbrightwell committed:
Added block crossover example

peterbrightwell committed:
Matrix -> mapping in 4.0

peterbrightwell committed:
Added overview terminology example/diagram

peterbrightwell committed:
Added overview example, added channel_index to block examples

peterbrightwell committed:
Removing duplicate diagrams following renaming

peterbrightwell committed:
Added channel_index to matrix example

peterbrightwell commented:
For the remaining TODO, will instead add two mapping matrices to show re-entrant behaviour.

peterbrightwell committed:
Updated links to map example images

garethsb-sony commented on docs/4.0. Behaviour.md:
```diff
@@ -102,60 +102,66 @@ Inputs may optionally constrain the way the channels contained in the Input may
 
-The following example describes a case, where the specific architecture of a given MADI Card allows only to either process Inputs 1-8, 9-16, 17-24, and so on from a given MADI Signal. The following output would be permissible, where a `block_size` value of `8` is used:
-->> These blocks are inconsistent with the requirement that each block must start with an index equal to a multiple of the block size. How to rectify?
+Consider a case where the specific architecture of a given MADI Card only allows processing of Input channels in blocks. The MADI channel numbers and corresponding Input `channel_index` values of these blocks are `0-7`, or `8-15`, or `16-23`, and so on.
```
```suggestion
Consider a case where the specific architecture of a given MADI Card only allows processing of Input channels in blocks. The MADI channel numbers and corresponding Input `channel_index` values of these blocks are `0`-`7`, or `8`-`15`, or `16`-`23`, and so on.
```

Also, just confirming, do **MADI channel numbers** start from zero?

peterbrightwell committed:
Text and diagram clarifcations, SHALL->MUST

peterbrightwell commented:
@garethsb-sony @chris-gil:

- Added the double-map diagram we discussed.
- All SHALLs are now MUSTs.
- Simplified the reordering text (@simonrankine you might want to check it's not too simplified now, as what were two bullet points have been combined).
- Added "different offsets" visual aid to the reordering example with gaps (hopefully the crossover one doesn't need it).

garethsb-sony commented:
@peterbrightwell do you think we should wordsmith https://github.com/AMWA-TV/nmos-audio-channel-mapping/blob/gh-pages/index.md slightly now that we fixed on 'mapping' rather than 'matrix'?

garethsb commented on docs/1.0. Overview.md:
```diff
@@ -31,6 +31,54 @@ Inputs and Outputs both have a number of Channels available for use. An Input re
+
+Note that:
+
+- The eight Output A channels are routed as a block to the eight Input A channels. This could be due to hardware limitationso n those channels, which can be signalled as a constraint in IS-08 (see below).
```
```suggestion
- The eight Output A channels are routed as a block to the eight Input A channels. This could be due to hardware limitations on those channels, which can be signalled as a constraint in IS-08 (see below).
```

garethsb commented on docs/1.0. Overview.md:
```diff
@@ -31,6 +31,54 @@ Inputs and Outputs both have a number of Channels available for use. An Input re
+- Channels 5 and 6 of Output B are parked.
+
+
+The above Map is represented in JSON in the API as below. See [4.0 Behaviour](4.0.%20Behaviour.md) for more details.
```
```suggestion
The above Map is represented in JSON in the API as below. See [Behaviour](4.0.%20Behaviour.md) for more details.
```

(We've tended elsewhere to avoid section numbers in link titles; they don't appear in the rendered docs.)

garethsb commented on docs/4.0. Behaviour.md:
```diff
@@ -102,60 +102,61 @@ Inputs may optionally constrain the way the channels contained in the Input may
 
-The following example describes a case, where the specific architecture of a given MADI Card allows only to either process Inputs 1-8, 9-16, 17-24, and so on from a given MADI Signal. The following output would be permissible, where a `block_size` value of `8` is used:
-->> These blocks are inconsistent with the requirement that each block must start with an index equal to a multiple of the block size. How to rectify?
+Consider a case where the specific architecture of a given MADI Card only allows processing of Input channels in blocks. The MADI channel numbers and corresponding Input `channel_index` values of these blocks are `0-7`, or `8-15`, or `16-23`, and so on.
```
```suggestion
Consider a case where the specific architecture of a given MADI Card only allows processing of Input channels in blocks. The MADI channel numbers and corresponding Input `channel_index` values of these blocks are `0`-`7`, or `8`-`15`, or `16`-`23`, and so on.
```

Also, just confirming, do **MADI channel numbers** start from zero?

garethsb commented on docs/4.0. Behaviour.md:
```diff
@@ -244,22 +245,19 @@ A client cancels an activation by submitting a DELETE request to the resource `/
+All timestamps in this API MUST be string formatted timestamps (\<seconds\>:\<nanoseconds\>) indicating time (absolute or relative) for activation. For relative times, the time '0:0' MUST be considered to be the time the request is received by the API.
 
-In order to ensure consistent operation of the API during and around leap seconds the API SHALL use TAI for all absolute timestamps. TAI does not apply leap seconds, and is therefore an integer number of seconds different to UTC. The timestamps used in the API SHALL be based on the UNIX epoch, meaning 1970-01-01 00:00:00 TAI would be represented as '0:0'.
+In order to ensure consistent operation of the API during and around leap seconds the API MUST use TAI for all absolute timestamps. TAI does not apply leap seconds, and is therefore an integer number of seconds different to UTC. The timestamps used in the API MUST be based on the UNIX epoch, meaning 1970-01-01 00:00:00 TAI would be represented as '0:0'.
```
Are we certain about that statement? It's a critical point and I'm actually very surprised I can't find a thorough discussion of TAI timestamps in IS-04, IS-05 or the top-level NMOS repo.

See e.g. https://cr.yp.to/proto/utctai.html which says:

>What is the [UNIX] epoch? Originally it was defined as the beginning of 1970 GMT. GMT, Greenwich Mean Time, is a traditional term for the time zone in England. Unfortunately, it is ambiguous; it can refer to a variety of astronomical time scales.
>
> Arthur David Olson's popular time library uses an epoch of 1970-01-01 00:00:10 TAI.

In my understanding, that 10 reflects the difference between UTC and TAI at that point, since the UNIX epoch is usually referenced to UTC. Therefore, doesn't the timestamp '0:0' represent 1970-01-01 00:00:10 TAI?

(The offset between UTC and TAI is currently 37 seconds. 2020-01-01 00:00:00 UTC was equivalent to 2020-01-01 00:00:37 TAI. [IERS Bulletin C (leap second announcements)](https://www.iers.org/SharedDocs/News/EN/BulletinC.html) is updated 6 months in advance when further leap seconds are to be added, at most biannually at the end of June or December.)

peterbrightwell replied:
I think you're right. Shouldn't this be the SMPTE (or PTP) epoch? UNIX epoch isn't mentioned anywhere else in NMOS specs. FWIW the `Time Context` example in MS-04 has zero based on SMPTE epoch.

peterbrightwell replied:
@garethsb-sony Chris and I just discussed this and reckon the best thing is to delete the second two sentences.

peterbrightwell replied:
See PR #41

garethsb commented on docs/4.0. Behaviour.md:
```diff
@@ -244,22 +245,19 @@ A client cancels an activation by submitting a DELETE request to the resource `/
+![re-entrant mapping example](images/map-examples/re-entrant-example.png "Re-entrant mapping example")
 
-->> Bits of JSON corresponding to the example matrix?
+In this scenario the audio undergoes two distinct routing operations - the first from the MADI input to the MADI card, and then a second from the MADI card to the AES67 output. In order to allow for these kinds of restrictions the Audio Channel Mapping API supports re-entrant mappings, in which audio at an Output may be routed back to an Input. This is signaled to the control system by the same S ...
```
The updated diagram definitely clearly explains the "two distinct routing operations", but this does mean it doesn't _feel_ "re-entrant". Since the `map` API resource is a single thing, thinking about it, would it be better to show both operations in a single routing matrix - or if not, maybe to show how the two routing matrices are represented in one `map` JSON?

peterbrightwell replied:
Agreed, and have redrawn as a single routing matrix.

garethsb committed:
There aren't four activation modes. Try for breviloquence.

garethsb committed:
Commit suggestion

garethsb committed:
Commit suggestion

garethsb committed:
Commit suggestion

simonrankine commented:
> @garethsb-sony @chris-gil:
> 
>     * Added the double-map diagram we discussed.
> 
>     * All SHALLs are now MUSTs.
> 
>     * Simplified the reordering text (@simonrankine you might want to check it's not too simplified now, as what were your two bullet points have been combined).
> 
>     * Added "different offsets" visual aid to the reordering example with gaps (hopefully the crossover one doesn't need it).

I've taken a look @peterbrightwell and I think the same meaning is still conveyed where you've removed those two bullets. I think I was just labouring the point a bit as a lot of this was conceptually fairly new.

peterbrightwell committed:
Re-entrant map shown as single matrix

peterbrightwell commented:
Now just using a single routing matrix reprsentation of the re-entrant mapping

peterbrightwell committed:
Updated re-entrant diagram and text

garethsb commented on docs/1.0. Overview.md:
````diff
@@ -31,6 +31,54 @@ Inputs and Outputs both have a number of Channels available for use. An Input re
+```json
+"map":{
+  "Output A":{
+    "chan 1" :{
````
Utterly trivial niggle: whitespace is a bit inconsistent in this JSON example (probably is in examples too, since I think it came from there...). Normally I think we have `"prop": {` (rather than with any whitespace before the colon or none after). Also more blank lines than necessary after the example.

garethsb replied:
D'oh, we realised I missed the obvious mistake we made that the `"map"` properties should be output ids like `"outA"` (no spaces allowed), and the properties of that must be the stringized channel indexes, like `"0"`, etc. not channel names.

peterbrightwell replied:
@garethsb-sony that's now corrected, and I've updated all the mapping examples to show both input and output channel indexes. Or indices if you prefer.

peterbrightwell committed:
Corrected Overview example map, added output channel indexes

peterbrightwell commented:
@garethsb-sony I think all your suggestions are resolved, but could you check.

garethsb commented on docs/1.0. Overview.md:
```diff
@@ -45,40 +45,38 @@ Note that:
       "channel_index": 1
     },
-    ...
+    …
```
I think perhaps the red-backed ellipsis (here and repeated twice just below) looks more off-putting than the three separate dots?

peterbrightwell committed:
Making … more visible

chris-gil reviewed (approved):
Changes approved
